### ISSUE TITLE
[BUG] Error: ENOENT: no such file or directory, open '/var/task/node_modules/playwright-core/browsers.json' on Vercel with Next.js

### ISSUE THREAD

**Issue by BrunoQuaresma** [microsoft/playwright#5862]
After upgrade Playwright to 1.9.2 this error is happening on Vercel functions.

```
2021-03-16T19:57:07.806Z	undefined	ERROR	Error: ENOENT: no such file or directory, open '/var/task/node_modules/playwright-core/browsers.json'
    at Object.openSync (fs.js:462:3)
    at Object.readFileSync (fs.js:364:35)
    at new Registry (/var/task/node_modules/playwright-core/lib/utils/registry.js:213:54)
    at new Playwright (/var/task/node_modules/playwright-core/lib/server/playwright.js:46:23)
    at Object.createPlaywright (/var/task/node_modules/playwright-core/lib/server/playwright.js:59:12)
    at setupInProcess (/var/task/node_modules/playwright-core/lib/inprocess.js:24:37)
    at Object.<anonymous> (/var/task/node_modules/playwright-core/lib/inprocess.js:42:18)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/var/task/node_modules/playwright-core/browsers.json'
}
RequestId: fafc5e6c-b338-4ba2-90b0-a019fd78923b Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

**Comment by mxschmitt** [comment]
Seems like a weird caching issue on Vercel, because the file itself `browsers.json` is there in the latest version:

https://unpkg.com/playwright-core@1.9.2/browsers.json

which version was the last working version?

**Comment by jeadev** [comment]
@mxschmitt I have no idea.. I'm building an app and I've discovered this package yesterday. I'm quite confused on how to solve this issue I can't find anything online.
I tried to install the package at build-time in vercel but it doesn't work either..

**Comment by jeadev** [comment]
Do you have any suggestions on where I can host my NextJS app for it to work with playwright ? (other than Vercel of course)

**Comment by pavelfeldman** [comment]
Could you share your use case and why you use playwright-core? I'd recommend using playwright package only, unless you are reusing parts of our code for a custom implementation of some tool loosely related to Playwright.

**Comment by BrunoQuaresma** [comment]
The use case is: Use the playwright inside of a Vercel function. Actually, I'm using playwright + this package https://github.com/JupiterOne/playwright-aws-lambda.

**Comment by j2is** [comment]
Having the same issue on Vercel with the same package, any ideas how to resolve?

**Comment by tiivik** [comment]
Same issue here using Vercel following this guide https://playwright.tech/blog/generate-opengraph-images-using-playwright

```
"playwright-aws-lambda": "^0.6.0"
"playwright-core": "^1.10.0"
```
Vercel log:
```
2021-03-26T13:46:34.687Z	e33df87f-4bd6-43bd-9769-c9a3db2a8efe	ERROR	Error: ENOENT: no such file or directory, open '/var/task/node_modules/playwright-core/browsers.json'
    at Object.openSync (fs.js:476:3)
    at Object.readFileSync (fs.js:377:35)
    at new Registry (/var/task/node_modules/playwright-core/lib/utils/registry.js:215:54)
    at new Playwright (/var/task/node_modules/playwright-core/lib/server/playwright.js:46:23)
    at Object.createPlaywright (/var/task/node_modules/playwright-core/lib/server/playwright.js:59:12)
    at setupInProcess (/var/task/node_modules/playwright-core/lib/inprocess.js:24:37)
    at Object.<anonymous> (/var/task/node_modules/playwright-core/lib/inprocess.js:42:18)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/var/task/node_modules/playwright-core/browsers.json'
}
RequestId: e33df87f-4bd6-43bd-9769-c9a3db2a8efe Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

**Comment by dbredvick** [comment]
Also having the same issue here. Does a downgrade fix this for y'all?

**Comment by zephraph** [comment]
Hey folks, so I don't think this is a playwright issue. It's also likely a cache issue. When Vercel prepares serverless functions, it traces what dependencies/files are used and only includes those in the payload. What's likely happening here is that the tracing mechanism  does "see" this browser.json file as a dependency and thus doesn't upload it to the lambda. I wrote a little about the [tracing process](https://just-be.dev/posts/vercel-lambda-size-error) if you're interested. 

I fixed this in my PR above, but note that I'm not using a Next.js api function but rather a Vercel api function. That's necessary because you'll have to explicitly include the `browsers.json` file in vercel's config in order to get it to work (and you can't do that w/ next functions). Also, as mentioned in my linked blog post, Next server dependencies are all bundled together so by placing playwright (and therefore chromium) in your Next.js function you're risking pushing yourself over the 50MB function limit.

I'd recommend that this issue closed and transferred over to one of Vercel's repos, likely https://github.com/vercel/nft.

**Comment by pavelfeldman** [comment]
Closing as per above, please feel free to file a new one and link this bug if there are remaining questions / bugs.

**Comment by mxschmitt** [comment]
I did some more investigation into this issue. It seems [nft](https://github.com/vercel/nft) as @zephraph wrote is used to trace the files which are needed during the runtime does not detect that the `browsers.json` is needed.

See that line:

https://github.com/microsoft/playwright/blob/e81a3c5901e024f7c41edf023daad60c68ed093d/src/utils/registry.ts#L238

If it get replaced from `fs.readFileSync` to `require`, it works. I created the following demonstration PR on their side, so hopefully they fix it, otherwise in Playwright it can be changed from `readFileSync` with `JSON.parse` to `require` which workarounds it.

See here: https://github.com/vercel/nft/pull/184

**Comment by mxschmitt** [comment]
The issue should be fixed in the `npm install playwright-core@next` version and lands in version 1.11. I tested it and it's working for one of my projects.

cc @BrunoQuaresma @jeadev @j2is @tiivik @dbredvick @zephraph

**Comment by BrunoQuaresma** [comment]
@mxschmitt I'm still having the same error unfortunately.

**Comment by BrunoQuaresma** [comment]
The version 1.8.0 is the one that is working for me.

**Comment by mxschmitt** [comment]
Weird, seems like Vercel changed something on their ends, since the function which was working on my application is now not working anymore and I did not change anything.

**Comment by juice49** [comment]
It seems this has broken again in the last 24 hours. I have a Vercel project I deployed yesterday that this is working for. Today, if I redeploy the same code (using the `--force` flag), I am getting the `browsers.json` error.

```
Error: Cannot find module '/var/task/node_modules/playwright-core/browsers.json'
```

Still doing a bit more research to make sure...

**Comment by kapilgorve** [comment]
I feel like such a noob for posting this but I was able to fix this issue for myself. I guess they changed something internal on how they handle paths. 
I was importing  a file like this - 
`const script = fs.readFileSync("./image.js", "utf-8");`
Used the dirname path and got rid of  `.`
`const script = fs.readFileSync(__dirname + "/image.js", "utf-8");`
Now my serverless function is working again.

**Comment by juice49** [comment]
I think @mxschmitt is right—this stopped working after @vercel/nft 0.12.0 was released.

0.12.0 was released 8 days ago, and I would guess it got deployed to Vercel's infrastructure in the past few days, causing this issue to appear again for new deployments. @vercel/nft 0.11.2 does not have this issue.

I have created a [test case](https://github.com/juice49/vercel-nft-playwright-test) and opened vercel/nft#211.
